src/geode/geosciences/lso_input.cpp

```diff
@@ -0,0 +1,783 @@
+                solid_builder_->create_tetrahedron( vertices );
+                std::getline( file_, line_ );
+            } while( std::getline( file_, line_ )
+                     && geode::detail::string_starts_with( line_, "TETRA" ) );
+            solid_builder_->compute_polyhedron_adjacencies();
+        }
+
+        void read_surfaces()
+        {
+            if( !geode::detail::string_starts_with( line_, "MODEL" ) )
+            {
+                geode::detail::goto_keyword( file_, "MODEL" );
+            }
+            facet_id_ = solid_->facet_attribute_manager()
+                            .find_or_create_attribute< geode::VariableAttribute,
+                                geode::uuid >( "facet_id", default_id_ );
+            std::getline( file_, line_ );
+            while( geode::detail::string_starts_with( line_, "SURFACE" ) )
+            {
+                std::istringstream iss{ line_ };
+                std::string keyword;
+                iss >> keyword;
+                const auto h_id = builder_.add_horizon();
+                builder_.set_horizon_name(
+                    h_id, geode::detail::read_name( iss ) );
+                const auto& horizon = model_.horizon( h_id );
+                read_tfaces( horizon );
+            }
+        }
+
+        void read_tfaces( const geode::Horizon3D& horizon )
+        {
+            std::getline( file_, line_ );
+            while( geode::detail::string_starts_with( line_, "TFACE" ) )
+            {
+                const auto id =
+                    builder_.add_surface( geode::MeshFactory::default_impl(
+                        geode::TriangulatedSurface3D::type_name_static() ) );
+                const auto& surface = model_.surface( id );
+                builder_.add_surface_in_horizon( surface, horizon );
+                surfaces_.push_back( id );
+                surface_is_internal_[id] = false;
+                read_key_vertices();
+                read_triangles( id );
+            }
+        }
+
+        void read_triangles( const geode::uuid& surface_id )
+        {
+            absl::flat_hash_map< geode::index_t, geode::index_t >
+                vertex_mapping;
+            auto builder = builder_.surface_mesh_builder( surface_id );
+            const auto component_id =
+                model_.surface( surface_id ).component_id();
+            while( std::getline( file_, line_ )
+                   && geode::detail::string_starts_with( line_, "TRGL" ) )
+            {
+                const auto tokens = get_tokens();
+                std::array< geode::index_t, 3 > facet_vertices;
+                std::array< geode::index_t, 3 > vertices;
+                for( const auto i : geode::Range{ 3 } )
+                {
+                    geode::index_t value;
+                    auto ok = absl::SimpleAtoi( tokens[i + 1], &value );
+                    OPENGEODE_EXCEPTION(
+                        ok, "[LSOInput] Error while reading tetra" );
```

Comment:
I thinks putting 'while reading triangles' would be more explicit

Reply:
Bad copy/paste...